### PR TITLE
Improve watch pod to handle failed or evicted pod

### LINKED ISSUE #305: Evicted training tasks pods are never deleted
When a training task pod is evicted by kubernetes, the backend waits forever for the pod to either complete or fail.

However, since "Evicted" is not a failure condition that is currently supported in the backend:

- the pod stays in `Evicted` state forever
- the backend waits (for the pod to compelte) forever
- the training tasks (and possibly compute plan) stay in the `doing` state forever.

### PR DESCRIPTION
**Pull request by Kelvin-M**
## Description
Improve `watch_pod` to handle failed or evicted pod.
 
## Closes issue(s)
Fix #305 

## Companion PRs
None

## How to test / repro
None

## Screenshots / Trace
None

## Changes include
- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

## Checklist
- [x] I have tested this code
- [ ] I have updated the Readme

## Other comments
None

### CODE REVIEW

File: backend/substrapp/tasks/k8s_backend.py
Comment (AurelienGasser) [inline backend/substrapp/tasks/k8s_backend.py:108]: IMO we should immediately exit the function here because the POD will not recover even after multiple attempts: the time.sleep is 0.2 secs and if I interpret the [doc](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy) correctly the first restart of the pod will happen after 10 sec.

Comment (Kelvin-M) [reply backend/substrapp/tasks/k8s_backend.py:108]: Yes, thank you for digging in this documentation !

File: backend/substrapp/tasks/k8s_backend.py
Comment (samlesu) [inline backend/substrapp/tasks/k8s_backend.py:109]: Does it make sense to continue the execution inside the while loop if `finished` is set to true? At first sight it seems we would like to directly exit the while loop.

File: backend/substrapp/tasks/k8s_backend.py
Comment (samlesu) [inline backend/substrapp/tasks/k8s_backend.py:156]: What's the reason for not setting the `error` variable inside the `if api_response.status.phase == 'Failed' or (api_response.status.reason and 'Evicted' in api_response.status.reason):` condition ?

To do that, maybe we could move the if condition to the end of code inside the while loop?

File: backend/substrapp/tasks/k8s_backend.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/k8s_backend.py:109]: Yes, you're right we can early exit with a break or continue :) (as we set finished to true)

File: backend/substrapp/tasks/k8s_backend.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/k8s_backend.py]: I change the code to be simpler, so it's not necessary to have this merge code :)

**Review by samlesu — APPROVED** [review]
LGTM

@Kelvin-M It would be great to wait for another approval before merging.

File: backend/substrapp/tasks/k8s_backend.py
Comment (samlesu) [reply backend/substrapp/tasks/k8s_backend.py:113]: Thanks!

File: backend/substrapp/tasks/k8s_backend.py
Comment (samlesu) [reply backend/substrapp/tasks/k8s_backend.py]: Thanks!

**Review by AurelienGasser — APPROVED** [review]

**Pull request merged**
